src/modules-arch/module-descriptor.js

```diff
@@ -0,0 +1,221 @@
+import { store } from '@/vuex/index'
+import { vuexTypes } from '@/vuex/types'
+
+/**
+ * Describes the module. Its dependencies, incompatible modules,
+ * allowed submodules, component and store module importers.
+ */
+export class ModuleDescriptor {
+  /**
+   * Creates a new ModuleDescriptor.
+   *
+   * @param {Object} opts
+   *
+   * @param {Function} opts.importComponent
+   * Path to the entry point vue-component of the module. Please provide
+   * function like _ => import('./path).
+   *
+   * Neither require() nor import() work with variables. So you cannot do
+   * anything like require(SomeModule.componentPath). To make the components
+   * lazy-loaded we have to define them in a way like _ => import('./path)
+   * directly during the definition.
+   *
+   * @param {String} [opts.importStore]
+   * Path to the entry point vuex-module of the module. Please provide
+   * function like _ => import('./path).
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.dependencies]
+   * Constructors of modules the module depends on.
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.allowedSubmodules]
+   * Constructors of modules the module could use as submodules.
+   *
+   * @param {ModuleDescriptor[]} [opts.submodules]
+   * Modules that the module will use as submodules.
+   * Cannot contain any entries except those described in
+   * opts.allowedSubmodules.
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.incompatibles]
+   * Constructors of modules that cannot be used simultaneously with the module.
+   *
+   *
+   * @param {Boolean} [opts.isCorporateOnly]
+   * If `true` the component will be accessible only by corporate accounts
+   *
+   */
+  constructor (opts = {}) {
+    // will be assigned automatically while the component renders
+    this._createdComponentUid = -1
+    this._createdComponent = {}
+
+    const {
+      dependencies = [],
+      allowedSubmodules = [],
+      submodules = [],
+      incompatibles = [],
+      importComponent = null,
+      importStore = null,
+    } = opts
+
+    if (typeof importComponent !== 'function') {
+      throw new Error(`${this.constructor.name}: no importComponent provided`)
+    }
+
+    this._importComponent = this._rememberUidAfterImport(importComponent)
+
+    if (importStore) {
+      this._importStore = importStore
+    }
+
+    this._dependencies = dependencies
+    this._incompatibles = incompatibles
+    this._allowedSubmodules = allowedSubmodules
+    this._isCorporateOnly = opts.isCorporateOnly || false
+
+    this.validateSubmodules(submodules)
```

Comment:
why don't we also validate dependencies and compatibility here?

Reply:
because we do not know the whole set of modules on the constructor execution time